src/rust/container/lib.rs

```diff
@@ -0,0 +1,347 @@
+                .await
+                .map_err(|e| capnp::Error::failed(e.to_string()))?;
+
+            docker
+                .start_container(&container_name, None::<StartContainerOptions>)
+                .await
+                .map_err(|e| capnp::Error::failed(e.to_string()))?;
+
+            Ok(())
+        })
+    }
+
+    fn status(
+        &mut self,
+        _: container::StatusParams,
+        mut results: container::StatusResults,
+    ) -> Promise<(), capnp::Error> {
+        let container_name = self.container_name.clone();
+        let docker = self.docker.clone();
+        Promise::from_future(async move {
+            let inspect = docker
+                .inspect_container(&container_name, None::<InspectContainerOptions>)
+                .await
+                .map_err(|e| capnp::Error::failed(e.to_string()))?;
+
+            let mut builder = results.get();
+            if let Some(state) = inspect.state {
+                builder.set_running(state.running.unwrap_or(false));
+            }
+            Ok(())
+        })
+    }
+
+    fn monitor(
+        &mut self,
+        _: container::MonitorParams,
+        _: container::MonitorResults,
+    ) -> Promise<(), capnp::Error> {
+        let container_name = self.container_name.clone();
+        let docker = self.docker.clone();
+        Promise::from_future(async move {
+            let options = WaitContainerOptionsBuilder::default()
+                .condition("not-running")
+                .build();
+
+            let mut stream = docker.wait_container(&container_name, Some(options));
+
+            if let Some(wait_result) = stream.next().await {
+                let status_code = wait_result
+                    .map_err(ContainerError::SocketError)?
+                    .status_code;
+                if status_code == 0 {
+                    return Ok(());
+                }
+
+                return Err(capnp::Error::failed(format!(
+                    "Container exited with unexpected exit code: {status_code}"
+                )));
+            }
+
+            // Stream ended without result - should not happen
+            Err(capnp::Error::failed(
+                "Monitoring stream ended unexpectedly".to_owned(),
+            ))
+        })
+    }
+
+    /// This method is intentionally not implemented right now.
+    fn listen_tcp(
+        &mut self,
+        _params: container::ListenTcpParams,
+        _results: container::ListenTcpResults,
+    ) -> Promise<(), capnp::Error> {
+        Promise::from_future(async move {
+            Err(capnp::Error::unimplemented(
+                "listen_tcp is not implemented".to_owned(),
+            ))
+        })
+    }
+
+    /// TODO(soon): Implement this.
+    fn get_tcp_port(
+        &mut self,
+        _params: container::GetTcpPortParams,
+        _results: container::GetTcpPortResults,
+    ) -> Promise<(), capnp::Error> {
+        Promise::from_future(async move {
```

Comment:
You need to implement a `rpc::Container::Port::Server` and return an object of that kind.
That object implements `connect` which pumps bytes back and forth between the container tcp port and the incoming bytestream.